File: 10_regular_expression_matching.py
"""
Leetcode: https://leetcode.com/problems/regular-expression-matching/
Approch: Dynamic Programming: Bottom-Up Variation
Time Complexity: O(T*P) where T is length of the text, P is the length of the pattern.
Space Complexity: O(T*P), the only memory we use is the O(T*P) boolean entries in our cache.
"""


class Solution:
    def isMatch(self, text: str, pattern: str) -> bool:
        dp = [[False] * (len(pattern) + 1) for _ in range(len(text) + 1)]

        dp[-1][-1] = True
        for i in range(len(text), -1, -1):
            for j in range(len(pattern) - 1, -1, -1):
                first_match = i < len(text) and pattern[j] in {text[i], '.'}
                if j+1 < len(pattern) and pattern[j+1] == '*':
                    dp[i][j] = dp[i][j+2] or first_match and dp[i+1][j]
                else:
                    dp[i][j] = first_match and dp[i+1][j+1]

        return dp[0][0]

File: 10-RegularExpressionMatching.cpp
/* 
    Time Complexity                              :  O(M*N) where M is the size of the string and N is the size of the pattern.
    Space Complexity                             :  O(M*N) size of the dp array.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size();
        int n = p.size();
        vector<vector<bool>> dp(m+1,vector<bool>(n+1, false));
        dp[0][0] = true;
        for(int j=1;j<=n;j++) {
            if(p[j-1] == '*') {
                dp[0][j] = dp[0][j-2];
            }
        }
        
        for(int i=1;i<=m;i++) {
            for(int j=1;j<=n;j++) {
                if(s[i-1] == p[j-1] || p[j-1] == '.') {
                    dp[i][j] = dp[i-1][j-1];
                } else if(p[j-1] == '*' ){
                    // 0 case
                    dp[i][j] = dp[i][j-2];
                    // 1 case
                    if(s[i-1] == p[j-2] || p[j-2] == '.') {
                        dp[i][j] = dp[i][j] || dp[i-1][j];
                    }
                } 
            }
        }
        
        
        return dp[m][n];
    }
};